Sources/SQLiteNIO/SQLiteConnection.swift

```diff
@@ -164,26 +164,13 @@ public final class SQLiteConnection: SQLiteDatabase {
         _ onRow: @escaping (SQLiteRow) -> Void
     ) -> EventLoopFuture<Void> {
         logger.debug("\(query) \(binds)")
-        let promise = self.eventLoop.makePromise(of: Void.self)
         return self.threadPool.runIfActive(eventLoop: self.eventLoop) {
-            do {
-                let statement = try SQLiteStatement(query: query, on: self)
-                try statement.bind(binds)
-                let columns = try statement.columns()
-                var callbacks: [EventLoopFuture<Void>] = []
-                while let row = try statement.nextRow(for: columns) {
-                    let callback = self.eventLoop.submit {
-                        onRow(row)
-                    }
-                    callbacks.append(callback)
-                }
-                EventLoopFuture<Void>.andAllSucceed(callbacks, on: self.eventLoop)
-                    .cascade(to: promise)
-            } catch {
-                promise.fail(error)
+            let statement = try SQLiteStatement(query: query, on: self)
+            try statement.bind(binds)
+            let columns = try statement.columns()
+            while let row = try statement.nextRow(for: columns) {
+                onRow(row)
```

Comment:
Doing it this way changes the behavior so that the `onRow` callback is invoked on the thread pool's runner thread rather than the event loop's thread; it would probably be better that way in an ideal world, but unfortunately it's likely to cause hard-to-understand (not to mention debug) problems for anyone who happens to rely on the existing behavior (which is really easy to end up doing by accident). This is an atypical case where we actually _do_ need to know whether or not the thread pool is active before acting.

The best way to do this is undoing the quite frankly extremely lazy changes I suggested in the custom functions PR (yep, I caused the bug by insisting on doing things "the right way" and then not actually doing so 🤦‍♀️). Here's a revised implementation which also gets rid of a half dozen or so unnecessary allocations, and for good measure adds a little `Sendable` love by mimicking the `@preconcurrency` pattern NIO uses 😅.

(P.S.: If you happened to feel like adding the concurrency bit to the other various methods in this file which have closures in their parameter lists (`_SQLiteDatabaseCustomLogger.withConnection(_:)`, `_SQLiteDatabaseCustomLogger.query(_:_:logger:_:)`, and `SQLiteConnection.withConnection(_:)`), you're very very welcome to do so - I won't reject the PR if you don't, though! 😆)

```swift
public protocol SQLiteDatabase {
    // <snip>

    #if swift(>=5.7)
    @preconcurrency func query(
        _ query: String,
        _ binds: [SQLiteData],
        logger: Logger,
        _ onRow: @escaping @Sendable (SQLiteRow) -> Void
    ) -> EventLoopFuture<Void>
    #else
    func query(
        _ query: String,
        _ binds: [SQLiteData],
        logger: Logger,
        _ onRow: @escaping (SQLiteRow) -> Void
    ) -> EventLoopFuture<Void>
    #endif

    // <snip>
}

// <snip>

public final class SQLiteConnection: SQLiteDatabase {
    // <snip>

    #if swift(>=5.7)
    @preconcurrency public func query(
        _ query: String, _ binds: [SQLiteData], logger: Logger,
        _ onRow: @escaping @Sendable (SQLiteRow) -> Void
    ) -> EventLoopFuture<Void> {
        self._query(query, binds, logger: logger, onRow)
    }
    #else
    public func query(
        _ query: String, _ binds: [SQLiteData], logger: Logger,
        _ onRow: @escaping (SQLiteRow) -> Void
    ) -> EventLoopFuture<Void> {
        self._query(query, binds, logger: logger, onRow)
    }
    #endif
    
    private func _query(
        _ query: String, _ binds: [SQLiteData], logger: Logger, _ onRow: @escaping (SQLiteRow) -> Void
    ) -> EventLoopFuture<Void> {
        logger.debug("\(query) \(binds)")
        let promise = self.eventLoop.makePromise(of: Void.self)
        self.threadPool.submit {
            guard case $0 = NIOThreadPool.WorkItemState.active else {
                // Note: We should be throwing NIOThreadPoolError.ThreadPoolInactive here, but we can't
                // 'cause its initializer isn't public so we let `SQLITE_MISUSE` get the point across.
                return promise.fail(SQLiteError(reason: .misuse, message: "Thread pool is inactive"))
            }
            var futures: [EventLoopFuture<Void>] = []
            do {
                let statement = try SQLiteStatement(query: query, on: self)
                let colums = try statement.columns()
                try statement.bind(binds)
                while let row = try statement.nextRow(for: columns) {
                    futures.append(self.eventLoop.submit { onRow(row) })
                }
            } catch {
                return promise.fail(error) // EventLoopPromise.fail(_:), conveniently, returns Void
            }
            EventLoopFuture.andAllSucceed(futures, promise: promise)
        }
        return promise.futureResult
    }
```

Reply:
🤣

I only noticed after I'd finished writing all this up that the failed `dependents-check` test was caused by this exact concern triggering an assertion in FluentKit 😅

Reply:
Great suggestions! I've changed the query implementation over to the one you provided above, and added Swift 5.7 Sendable support to the functions you listed.

Thanks for the review!